fsspec/implementations/reference.py

```diff
@@ -172,11 +199,14 @@ def __init__(
         if remote_protocol is None:
             remote_protocol = target_protocol
 
-        self.fs = filesystem(remote_protocol, loop=loop, **(remote_options or {}))
+        fs = fs or filesystem(remote_protocol, loop=loop, **(remote_options or {}))
+        self.fss[remote_protocol] = fs
```

Comment:
What if there are multiple filesystems with the same protocol but different locations? E.g. s3 with different endpoints, ssh with different hosts,etc ?

Seems like ideally something like https://github.com/fsspec/filesystem_spec/issues/773 should be used here instead.

Reply:
You are right, this is a problem, and a way of referring to unique combinations of arguments is very useful. But even if we can specify these filesystems, how would the individual references point to them? They have URLs like "s3://bucket/.." .

Reply:
I am thinking perhaps we could move to basing the FS choice on a *prefix* rather than protocol, so that "s3://bucket1/..." might be a different endpoint to "s3://bucket2/..." . This seems achievable, but also complex for a user to understand and wield. Perhaps that doesn't matter is the details of the options for each filesystem is not really something the user gets to see.

Reply:
I am settling for protocols only for now, but the door is open to have FSs per prefix, if we can avoid expensive string matching (.startswith) at access time.